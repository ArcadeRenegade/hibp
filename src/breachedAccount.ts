import fetchFromApi from './internal/haveibeenpwned/fetchFromApi';
import { Breach } from './types/remote-api.d';

/**
 * Fetches breach data for a specific account.
 *
 * ***Warning:***
 *
 * As of January, 2019, `haveibeenpwned.com` has started blocking requests to
 * the `breachedaccount` endpoint when originating from within a browser (based
 * on the `User-Agent` field of the request headers). To use this function in a
 * browser, you will likely have to proxy your request through a server of your
 * own. The `baseUrl` option was added to facilitate this workaround.
 *
 * @param {string} account a username or email address
 * @param {object} [options] a configuration object
 * @param {string} [options.domain] a domain by which to filter the results
 * (default: all domains)
 * @param {boolean} [options.includeUnverified] include "unverified" breaches in
 * the results (default: true)
 * @param {boolean} [options.truncate] truncate the results to only include
 * the name of each breach (default: true)
 * @param {string} [options.baseUrl] a custom base URL for the
 * haveibeenpwned.com API endpoints (default:
 * `https://haveibeenpwned.com/api/v3`)
 * @param {string} [options.userAgent] a custom string to send as the User-Agent
 * field in the request headers (default: `hibp <version>`)
 * @returns {(Promise<Breach[]> | Promise<null>)} a Promise which resolves to an
 * array of breach objects (or null if no breaches were found), or rejects with
 * an Error
 * @example
 * breachedAccount('foo')
 *   .then(data => {
 *     if (data) {
 *       // ...
 *     } else {
 *       // ...
 *     }
 *   })
 *   .catch(err => {
 *     // ...
 *   });
 * @example
 * breachedAccount('bar', {
 *   includeUnverified: false,
 *   baseUrl: 'https://my-hibp-proxy:8080',
 * })
 *   .then(data => {
 *     if (data) {
 *       // ...
 *     } else {
 *       // ...
 *     }
 *   })
 *   .catch(err => {
 *     // ...
 *   });
 * @example
 * breachedAccount('baz', {
 *   domain: 'adobe.com',
 *   truncate: false,
 *   userAgent: 'my-app 1.0'
 * })
 *   .then(data => {
 *     if (data) {
 *       // ...
 *     } else {
 *       // ...
 *     }
 *   })
 *   .catch(err => {
 *     // ...
 *   });
 * @alias module:breachedAccount
 */
const breachedAccount = (
  account: string,
  options: {
    domain?: string;
    includeUnverified?: boolean;
    truncate?: boolean;
    baseUrl?: string;
    userAgent?: string;
  } = {
    includeUnverified: true,
    truncate: true,
  },
): Promise<Breach[] | null> => {
  const endpoint = `/breachedaccount/${encodeURIComponent(account)}?`;
  const params = [];
  if (options.domain) {
    params.push(`domain=${encodeURIComponent(options.domain)}`);
  }
  if (options.includeUnverified === false) {
    params.push('includeUnverified=false');
  }
  if (options.truncate === false) {
    params.push('truncateResponse=false');
  }
  return fetchFromApi(`${endpoint}${params.join('&')}`, {
    baseUrl: options.baseUrl,
    userAgent: options.userAgent,
  }) as Promise<Breach[] | null>;
};

/**
 * A module for retrieving breach data for a specific account.
 *
 * @module breachedAccount
 * @example
 * import { breachedAccount } from 'hibp';
 */
export default breachedAccount;
