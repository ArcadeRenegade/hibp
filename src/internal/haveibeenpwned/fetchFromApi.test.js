import {
  ERR,
  INVALID_HEADER,
  RATE_LIMITED,
  EMAIL_INVALID,
  UNEXPECTED,
} from '../../../test/fixtures';
import dataClasses from '../../dataClasses';
import breachedAccount from '../../breachedAccount';
import axiosInstance from './axiosInstance';

describe('internal (haveibeenpwned): fetchFromApi', () => {
  describe('request failure', () => {
    let failboat;

    beforeAll(() => {
      failboat = axiosInstance.interceptors.request.use(() => {
        throw ERR;
      });
    });

    afterAll(() => {
      axiosInstance.interceptors.request.eject(failboat);
    });

    it('should re-throw request setup errors', () =>
      expect(dataClasses()).rejects.toEqual(ERR));
  });

  describe('invalid account format', () => {
    it('should throw an Error with "Bad Request" status text', () =>
      expect(breachedAccount(EMAIL_INVALID)).rejects.toMatchSnapshot());
  });

  describe('invalid request header', () => {
    it('should throw an Error with "Forbidden" status text', () =>
      expect(breachedAccount(INVALID_HEADER)).rejects.toMatchSnapshot());
  });

  describe('rate limited', () => {
    it('should throw an Error with "Too Many Requests" response data', () =>
      expect(breachedAccount(RATE_LIMITED)).rejects.toMatchSnapshot());
  });

  describe('unexpected HTTP error', () => {
    it('should throw an Error with the response status text', () =>
      expect(breachedAccount(UNEXPECTED)).rejects.toMatchSnapshot());
  });
});
