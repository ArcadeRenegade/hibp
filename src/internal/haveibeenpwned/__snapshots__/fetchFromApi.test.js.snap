// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`internal (haveibeenpwned): fetchFromApi invalid account format should throw an Error with "Bad Request" status text 1`] = `[Error: Bad request â€” the account does not comply with an acceptable format.]`;

exports[`internal (haveibeenpwned): fetchFromApi invalid request header should throw an Error with "Forbidden" status text 1`] = `[Error: Forbidden - no user agent has been specified in the request.]`;

exports[`internal (haveibeenpwned): fetchFromApi rate limited should throw an Error with "Too Many Requests" response data 1`] = `[Error: Rate limit exceeded, refer to acceptable use of the API: https://haveibeenpwned.com/API/v2#AcceptableUse]`;

exports[`internal (haveibeenpwned): fetchFromApi unexpected HTTP error should throw an Error with the response status text 1`] = `[Error: Unknown - something unexpected happened.]`;
